### PR TITLE
Nano-transcriber / flow estimates

### PR DESCRIPTION
**Pull request by Riteshlogin**
Depends on a pull request in iarc7_msgs

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -13,8 +13,7 @@ find_package(catkin REQUIRED COMPONENTS
   iarc7_safety
   message_filters
   roscpp
-  tf
```
Comment (amiller27) [inline CMakeLists.txt]: Why were these removed?

File: CMakeLists.txt [hunk 2]
```diff
@@ -161,6 +160,29 @@ target_link_libraries(iarc7_sensors_altimeter_from_subscription_node
   ${catkin_LIBRARIES}
 )
 
+add_executable(iarc7_nano_transcriber
+   src/Nano_transcriber.cpp
+)
+ add_dependencies(iarc7_sensors_altimeter_from_subscription_node
```
Comment (amiller27) [inline CMakeLists.txt]: You need to change `iarc7_sensors_altimeter_from_subscription_node` here to `iarc7_nano_transcriber`, same below on line 178 with `iarc7_flow_transformer`

File: include/iarc7_sensors/FlowTransformer.hpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <geometry_msgs/TwistWithCovarianceStamped.h>
+
+#include <iarc7_msgs/FlowVector.h>
+
+namespace iarc7_sensors {
+
+struct FlowTransformerSettings {
+    double fov; // 0.73 rad
```
Comment (amiller27) [inline include/iarc7_sensors/FlowTransformer.hpp]: Can we move these comments to wherever these numbers are actually defined?

File: include/iarc7_sensors/FlowTransformer.hpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    double min_estimation_altitude; // Future reference - this is 80 mm - some transform tree value
+    double vertical_threshold; //
+    double variance;
+    double variance_scale;
+    double tf_timeout;
+    int image_width;
+    bool debug_print;
+};
+
+// variables - current / last orientation, target_size, current_altitude
```
Comment (amiller27) [inline include/iarc7_sensors/FlowTransformer.hpp]: What's this comment?

File: include/iarc7_sensors/FlowTransformer.hpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+class FlowTransformer {
+
```
Comment (amiller27) [inline include/iarc7_sensors/FlowTransformer.hpp]: Can you get rid of the extra whitespace in this file and clean up the formatting?

File: include/iarc7_sensors/FlowTransformer.hpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+    double current_altitude_;
+    tf2::Quaternion current_orientation_;
+    tf2::Quaternion last_orientation_;
+    geometry_msgs::TransformStamped current_camera_to_level_quad_tf_;
+    geometry_msgs::TransformStamped last_camera_to_level_quad_tf_;
+    const ros_utils::SafeTransformWrapper transform_wrapper_;
+
+    iarc7_sensors::FlowTransformerSettings flow_transformer_settings;
```
Comment (amiller27) [inline include/iarc7_sensors/FlowTransformer.hpp]: Need an underscore on the end of the variable name here

File: include/iarc7_sensors/FlowTransformer.hpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    const ros::Publisher debug_unrotated_vel_pub_;
+
+
+    // Filter that gets applied to x/y velocity calculations to prevent random zero
+    // crossings that should not exist
+    //
+    // velocity_filtered[0] has the x measurements
+    // [1] has the y measurements 
+    const float filter_coefs_[4] = {0.0677, 0.4323, 0.4323, 0.0677};
+    //const float filter_coefs_[5] = {0.0338, 0.2401, 0.4521, 0.2401, 0.0338};
```
Comment (amiller27) [inline include/iarc7_sensors/FlowTransformer.hpp]: What are these commented out variables here?

File: launch/altimeter.launch [hunk 8]
```diff
@@ -29,7 +29,7 @@
             <rosparam command="load"
                 file="$(find iarc7_sensors)/param/tfmini_altimeter.yaml" />
             <remap from="altimeter_reading" to="long_distance_lidar" />
-            <remap from="altimeter_pose" to="long_distance_lidar_pose" />
+            <remap from="altimeter_pose" to="altimeter_pose" />
```
Comment (amiller27) [inline launch/altimeter.launch]: If this is just a remap to itself you can delete the remap

File: launch/nano_subscriber.launch [hunk 9]
```diff
@@ -0,0 +1,8 @@
+<launch>
+
+     <node name="Nano_transcriber" pkg="iarc7_sensors"
```
Comment (amiller27) [inline launch/nano_subscriber.launch]: Nano_transcriber should be all lower case

File: param/flow_transformer.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+# Min variance to send with velocity measurements
+##variance: 0.25
+variance: 5.0
+
+#Min altitude to run optical flow
+vertical_threshold : 0.8
+
+#Activate debug prints
+debug_print : true
```
Comment (amiller27) [inline param/flow_transformer.yaml]: Should this be on by default?

File: launch/altimeter.launch (hunk 8 above)
Comment (asaba96) [inline launch/altimeter.launch]: Why are we remapping the topic to the same name as the topic?

File: param/flow_transformer.yaml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+# Timeout for updates from tf
+tf_timeout: 1.0
+
+# Number of pixels in x and y direction across half the "screen"
+image_width: 30
+
+# What to scale the variance on according to the rotation rate / probably needs to be changed
+variance_scale: 0.19 # 0.3s/90deg
+
+# Min variance to send with velocity measurements
```
Comment (asaba96) [inline param/flow_transformer.yaml]: Can the comments in this file be cleaned up please?

File: src/FlowTransformer.cpp [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+    double last_yaw, last_pitch, last_roll;
+    getYPR(last_orientation_, last_yaw, last_pitch, last_roll);
+
+    // Calculate time between last and current frame
+    double dt = (time - last_message_time_).toSec();
+    
+    //ROS_ERROR("time is %f\n", time.toSec());
+    //ROS_ERROR("last time is %f\n", last_message_time_.toSec());
+
+    //ROS_ERROR("dt is %lf\n", dt);
```
Comment (asaba96) [inline src/FlowTransformer.cpp]: Can the commented out print statements be cleaned up?

File: src/FlowTransformer.cpp [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+    //ROS_ERROR_STREAM("Filtered X Vel: " << filteredXVel << " estimated X Vel: " << estimatedXVel);
+    //ROS_ERROR_STREAM("Filtered Y Vel: " << filteredYVel << " estimated Y Vel: " << estimatedYVel);
+
+    //filteredXVel = estimatedXVel;
+    //filteredYVel = estimatedYVel;
+
+    /*double dp;
+    double dr;
+
+    // These two if statements make sure that dp and dr are the shortest change
```
Comment (asaba96) [inline src/FlowTransformer.cpp]: Half of this file is commented out. Is the code not needed?

**Review by asaba96 — COMMENTED** [review]
Picky issues so far. Will look over again in more depth.

File: src/FlowTransformer.cpp (hunk 13 above)
Comment (Riteshlogin) [reply src/FlowTransformer.cpp]: I am seeing some questionable things I left behind. I will remove them when I get a chance

File: src/FlowTransformer.cpp [hunk 14]
```diff
@@ -0,0 +1,459 @@
+#include "iarc7_sensors/FlowTransformer.hpp"
+
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <tf2/LinearMath/Matrix3x3.h>
+#include <tf2/LinearMath/Quaternion.h>
+
+#include <iarc7_msgs/FlowVector.h>
+#include <math.h>
```
Comment (amiller27) [inline src/FlowTransformer.cpp]: Use `#include <cmath>`

File: src/FlowTransformer.cpp [hunk 15]
```diff
@@ -0,0 +1,459 @@
...
+                                 callback);
+
+    ros::spin();
+
+}
+
+
+namespace iarc7_sensors{
+
+FlowTransformer::FlowTransformer(
```
Comment (amiller27) [inline src/FlowTransformer.cpp]: Fix the formatting here

Comment (amiller27) [reply src/FlowTransformer.cpp]: Match the formatting in `iarc7_motion/src/QuadVelocityController.cpp`

File: src/FlowTransformer.cpp [hunk 16]
```diff
@@ -0,0 +1,459 @@
...
+                    nh.advertise<geometry_msgs::TwistWithCovarianceStamped>(
+                    "twist_correction", 10)),
+                debug_raw_pub_(
+                    nh.advertise<geometry_msgs::TwistWithCovarianceStamped>(
+                    "twist_raw", 10)),
+                debug_unrotated_vel_pub_(
+                    nh.advertise<geometry_msgs::TwistWithCovarianceStamped>(
+                    "twist_unrotated", 10))
+{
+    return;
```
Comment (amiller27) [inline src/FlowTransformer.cpp]: You don't need this

File: src/Nano_transcriber.cpp [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+
+    sensor_msgs::Range long_range_msg;
+    sensor_msgs::Range short_range_msg;
+    iarc7_msgs::FlowVector flow_msg;
+    iarc7_msgs::Float64Stamped battery_voltage;
+
+    long_range_msg.range = nano_info.long_range;
+    long_range_msg.min_range = 0.5;
+    long_range_msg.max_range = 10;
+    ros::Duration long_range_offset = ros::Duration().fromNSec((int64_t)(nano_info.long_range_offset * 10000));
```
Comment (amiller27) [inline src/Nano_transcriber.cpp]: Why the 10000?

File: src/Nano_transcriber.cpp [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+void sendESCCommand(iarc7_msgs::PlanarThrottle esc_cmnds,
+                    ros::Publisher& esc_cmd_pub,
+                    bool activateSideRotors)
+{
+
+    iarc7_msgs::ESCCommand esc_cmd;
+
+    if(activateSideRotors)
+    {
+        esc_cmd.front_motor_PWM = (esc_cmnds.front_throttle*125) + 125;
```
Comment (amiller27) [inline src/Nano_transcriber.cpp]: Why the 125?

File: src/Nano_transcriber.cpp [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+    return;
+
+}
+
+int main(int argc, char* argv[]){
+    
+    ros::init(argc, argv, "Nano_transcriber");
+    ros::NodeHandle nh;
+
+    ROS_WARN("Nano transcriber beginning");
```
Comment (amiller27) [inline src/Nano_transcriber.cpp]: Make this a `ROS_INFO` or `ROS_DEBUG`

File: src/Nano_transcriber.cpp (hunk 18 above)
Comment (Riteshlogin) [reply src/Nano_transcriber.cpp]: This is to turn a number from 0-1 into a corresponding number in the int8 range, which corresponds to the duty cycle sent to each side rotor.

File: src/Nano_transcriber.cpp (hunk 19 above)
Comment (Riteshlogin) [reply src/Nano_transcriber.cpp]: Put that there when I couldn't get this to run at all and have not had that problem since I fixed it, do you care if I just delete it

File: src/Nano_transcriber.cpp (hunk 17 above)
Comment (Riteshlogin) [reply src/Nano_transcriber.cpp]: We get a time offset in microseconds from the atmega, so this is to turn microseconds to nanoseconds. I will add a corresponding comment

File: include/iarc7_sensors/FlowTransformer.hpp (hunk 4 above)
Comment (Riteshlogin) [reply include/iarc7_sensors/FlowTransformer.hpp]: Something that should not be there

File: src/Nano_transcriber.cpp (hunk 18 above)
Comment (amiller27) [reply src/Nano_transcriber.cpp]: Please add a comment

File: src/Nano_transcriber.cpp (hunk 19 above)
Comment (amiller27) [reply src/Nano_transcriber.cpp]: Yeah just delete it

**Comment by amiller27** [comment]
Also this needs to get rebased

**Comment by asaba96** [comment]
EXTREMELY PICKY but can the name of the nano transcriber file and class follow the other conventions, so NanoTranscriber.cpp instead of Nano_transcriber.cpp

File: include/iarc7_sensors/FlowTransformer.hpp (hunk 7 above)
Comment (Riteshlogin) [reply include/iarc7_sensors/FlowTransformer.hpp]: Filters that were experimented with but that we decided didn't work as well

File: param/flow_transformer.yaml (hunk 10 above)
Comment (Riteshlogin) [reply param/flow_transformer.yaml]: So it says "debug_print", but really it decides whether things get published to debug topics. Do you still think it should be off by default?

File: src/FlowTransformer.cpp [hunk 20]
```diff
@@ -275,19 +213,6 @@ FlowTransformer::estimateVelocityFromFlowVector(const int deltaX, const int delt
...
-    geometry_msgs::PointStamped curr_pos, last_pos;
-    tf2::doTransform(curr_pos, curr_pos, current_camera_to_level_quad_tf_);
-    tf2::doTransform(last_pos, last_pos, last_camera_to_level_quad_tf_);
-    double camera_relative_vel_x = (curr_pos.point.x - last_pos.point.x) / dt;
-    double camera_relative_vel_y = (curr_pos.point.y - last_pos.point.y) / dt;
-    */
-
     Eigen::Vector3d corrected_level_quad_vel = level_quad_vel;
     //corrected_level_quad_vel.x() -= camera_relative_vel_x;
     //corrected_level_quad_vel.y() -= camera_relative_vel_y;
```
Comment (amiller27) [inline src/FlowTransformer.cpp]: These two lines need to get deleted too

File: src/FlowTransformer.cpp [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+        return;
+    }
+
+
+    const geometry_msgs::TwistWithCovarianceStamped velocity
+        = estimateVelocityFromFlowVector(flow_vector.deltaX, 
+        flow_vector.deltaY,
+        flow_vector.header.stamp);
+
+    //ROS_INFO("Before checking whether we have some infinite numbers");
```
Comment (amiller27) [inline src/FlowTransformer.cpp]: Get rid of this line

File: src/FlowTransformer.cpp [hunk 22]
```diff
@@ -5,7 +5,7 @@
 #include <tf2/LinearMath/Quaternion.h>
```
Comment (amiller27) [inline src/FlowTransformer.cpp]: Can you delete the empty duplicate lines in this file?

File: src/NanoTranscriber.cpp [hunk 23]
```diff
@@ -15,6 +15,9 @@ void distributeMessages(iarc7_msgs::Nano nano_info,
                         ros::Publisher& long_range_pub,
                         ros::Publisher& battery_publisher){
 
+    // From here on out, each of the "offset" variables is multiplied by 1000 to convert
```
Comment (amiller27) [inline src/NanoTranscriber.cpp]: You mean 10000?  And tens of microseconds?

File: include/iarc7_sensors/FlowTransformer.hpp (hunk 3 above)
Comment (amiller27) [reply include/iarc7_sensors/FlowTransformer.hpp]: This wasn't done

File: include/iarc7_sensors/FlowTransformer.hpp (hunk 4 above)
Comment (amiller27) [reply include/iarc7_sensors/FlowTransformer.hpp]: This wasn't removed

File: param/flow_transformer.yaml (hunk 10 above)
Comment (amiller27) [reply param/flow_transformer.yaml]: Yes

File: src/NanoTranscriber.cpp (hunk 23 above)
Comment (Riteshlogin) [reply src/NanoTranscriber.cpp]: They should be 1000, that was a mistake

Comment (aftersomemath) [reply src/NanoTranscriber.cpp]: I think its supposed to be 10000

Comment (Riteshlogin) [reply src/NanoTranscriber.cpp]: Yes, there was a small modification in the firmware that I was unaware of

**Review by asaba96 — COMMENTED** [review]
Looks good to me

**Comment by asaba96** [comment]
If no one has anymore concerns with this, I will merge it.

**Comment by aftersomemath** [comment]
Usually you shouldn't merge until all reviewers have confirmed that all issues they raised are resolved

File: launch/altimeter.launch [hunk 24]
```diff
@@ -29,7 +29,6 @@
             <rosparam command="load"
                 file="$(find iarc7_sensors)/param/tfmini_altimeter.yaml" />
             <remap from="altimeter_reading" to="long_distance_lidar" />
```
Comment (amiller27) [inline launch/altimeter.launch]: This line should be deleted as well if the line below is deleted

File: src/FlowTransformer.cpp [hunk 25]
```diff
@@ -0,0 +1,360 @@
...
+    ROS_ASSERT(private_nh.getParam(
+        "debug_print",
+        settings.debug_print));
+
+}
+
+
+int main(int argc, char* argv[]){
+
+    ros::init(argc, argv, "optical_flow_estimator");
```
Comment (amiller27) [inline src/FlowTransformer.cpp]: This should be `flow_transformer`, `optical_flow_estimator` is ambiguous

**Comment by asaba96** [comment]
Of course but sometimes people forget to review so I made that comment in the hope reviewers would review.

**Comment by amiller27** [comment]
It's been 17 hours since our last activity, and for most of that we were asleep.  Give me a couple minutes, I have a few more comments.

File: src/FlowTransformer.cpp [hunk 26]
```diff
@@ -0,0 +1,360 @@
...
+    double current_meters_per_px = distance_to_plane * std::tan(flow_transformer_settings_.fov/2)
+                                    / (flow_transformer_settings_.image_width/2.0);
+
+
+    // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    // These velocities are flipped, but it really should be implemented properly in the 
+    // transform tree. Definitely fix this later.
+    // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+    float estimatedXVel = current_meters_per_px * deltaX / std::cos(pitch)/ dt;
```
Comment (amiller27) [inline src/FlowTransformer.cpp]: Please change all floats to doubles

File: src/FlowTransformer.cpp [hunk 27]
```diff
@@ -0,0 +1,360 @@
...
+
+    last_message_time_ = flow_vector.header.stamp;
+
+    if (!std::isfinite(velocity.twist.twist.linear.x)
+     || !std::isfinite(velocity.twist.twist.linear.y)
+     || !std::isfinite(velocity.twist.covariance[0])    // variance of x
+     || !std::isfinite(velocity.twist.covariance[1])    // covariance of x & y
+     || !std::isfinite(velocity.twist.covariance[6])    // covariance of x & y
+     || !std::isfinite(velocity.twist.covariance[7])) { // variance of y
+        ROS_ERROR_STREAM("Invalid measurement in OpticalFlowEstimator: "
```
Comment (amiller27) [inline src/FlowTransformer.cpp]: This needs to say `FlowTransformer`, not `OpticalFlowEstimator`

File: src/FlowTransformer.cpp [hunk 28]
```diff
@@ -0,0 +1,360 @@
...
+    last_message_time_ = flow_vector.header.stamp;
+    last_orientation_ = current_orientation_;
+    last_camera_to_level_quad_tf_ = current_camera_to_level_quad_tf_;
+}
+
+bool FlowTransformer::canEstimateFlow()
+{
+    if (current_altitude_ < flow_transformer_settings_.min_estimation_altitude) {
+        ROS_WARN_THROTTLE(2.0,
+                          "Optical flow: height (%f) is below min processing height (%f)",
```
Comment (amiller27) [inline src/FlowTransformer.cpp]: Error message needs to say flow transformer, not optical flow

**Comment by asaba96** [comment]
😁

File: src/NanoTranscriber.cpp [hunk 29]
```diff
@@ -0,0 +1,171 @@
...
+
+    sensor_msgs::Range long_range_msg;
+    sensor_msgs::Range short_range_msg;
+    iarc7_msgs::FlowVector flow_msg;
+    iarc7_msgs::Float64Stamped battery_voltage;
+
+    long_range_msg.range = nano_info.long_range;
+    long_range_msg.min_range = 0.5;
+    long_range_msg.max_range = 10;
+    ros::Duration long_range_offset = ros::Duration().fromNSec((int64_t)(nano_info.long_range_offset * 10000));
```
Comment (amiller27) [inline src/NanoTranscriber.cpp]: Don't you want to cast `long_range_offset` before multiplying?

Comment (amiller27) [reply src/NanoTranscriber.cpp]: I believe the current code would promote `long_range_offset` to an `int` and then multiply, which would overflow for offsets larger than 2.14 seconds.

File: src/NanoTranscriber.cpp [hunk 30]
```diff
@@ -0,0 +1,171 @@
...
+    long_range_msg.header.stamp = nano_info.msg_received - long_range_offset;
+    long_range_msg.radiation_type = sensor_msgs::Range::INFRARED;
+    long_range_msg.field_of_view = 0.3;
+    long_range_msg.header.frame_id = "/tfmini";
+    if(nano_info.long_range_offset > 0) {
```
Comment (amiller27) [inline src/NanoTranscriber.cpp]: Does an offset of 0 mean that measurement is invalid?  If so can you put a comment saying that here?

File: src/NanoTranscriber.cpp [hunk 31]
```diff
@@ -0,0 +1,171 @@
...
+        planar_throttle.back_throttle = std::max(std::min(joy.axes[1], 1.0f), 0.0f);
+        planar_throttle.right_throttle = std::max(std::min(joy.axes[0], 1.0f), 0.0f);
+        planar_throttle.left_throttle = std::max(std::min(-joy.axes[0], 1.0f), 0.0f);
+
+        if(joystickActive) {
+            sendESCCommand(planar_throttle, esc_cmd_publisher, true);
+        }
+    };
+
+    // & just means "look at all the objects already declared and pull the ones you need"
```
Comment (amiller27) [inline src/NanoTranscriber.cpp]: This comment needs to disappear or go next to the lambda

File: src/NanoTranscriber.cpp (hunk 30 above)
Comment (Riteshlogin) [reply src/NanoTranscriber.cpp]: Yes, the offsets are set to 0 by default if the sensor is not ready with a measurement yet. I will add a comment there

**Comment by Riteshlogin** [comment]
Every issue should be resolved now. However, I have no way to test the altimeter.launch change currently and I am not 100% sure that it doesn't break anything.

**Review by amiller27 — APPROVED** [review]
